File: Data Structures & Algorithms/search-for-word/submission-0.cpp
class Solution {
public:

    set<pair<int, int>> visited;
    bool dfs(int row, int col, int idx, vector<vector<char>>& board, string word, int n, int m) {
        if (idx == word.size()) {
            return true;
        }
        if (row < 0 or row >= n or col < 0 or col >= m or board[row][col] != word[idx] or visited.count({row, col}) ) {
            return false;
        }
        bool result = false;
        visited.insert({row, col});
        result = dfs(row + 1, col, idx + 1, board, word, n, m) 
                or dfs(row - 1, col, idx + 1, board, word, n, m) 
                or dfs(row, col + 1, idx + 1, board, word, n, m)
                or dfs(row, col - 1, idx + 1, board, word, n, m);
        visited.erase({row, col});
        if (result) {
            return true;
        }
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {
        int n = board.size();
        int m = board[0].size();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (dfs(i, j, 0, board, word, n, m)) {
                    return true;
                }
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/search-for-word-ii/submission-0.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end = False

    def add(self, word):
        curr = self
        for ch in word:
            if ch not in curr.children:
                curr.children[ch] = TrieNode()
            curr = curr.children[ch]
        curr.end = True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        rows = len(board)
        cols = len(board[0])

        root = TrieNode()
        # build trie of the words
        for word in words:
            root.add(word)
        
        res, visited = set(), set()

        def dfs(r, c, node, word):
            if (r < 0 or c < 0 or r == rows or c == cols or (r, c) in visited or board[r][c] not in node.children):
                return
            
            node = node.children[board[r][c]]
            visited.add((r, c))
            word += board[r][c]

            if node.end:
                res.add(word)
            
            dfs(r + 1, c, node, word)
            dfs(r - 1, c, node, word)
            dfs(r, c + 1, node, word)
            dfs(r, c - 1, node, word)
            
            visited.remove((r, c))


        for row in range(rows):
            for col in range(cols):
                dfs(row, col, root, "")

        return list(res)